src/Data/Linear/Internal/MutableArray.hs

```diff
@@ -0,0 +1,46 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+-- |
+-- Module: MutableArray
+-- Description: Unsafe wrappers around `MutableArray#`s
+
+module Data.Linear.Internal.MutableArray
+  ( unsafeNewMutArr
+  , readMutArr
+  , writeMutArr
+  )
+where
+
+
+import GHC.Exts
+
+
+
+-- # Unsafe wrappers
+----------------------------
+
+unsafeNewMutArr :: Int -> a -> MutableArray# RealWorld a
+unsafeNewMutArr (I# size) init =
+  case newArray of
+    (# _, array #) -> array
+  where
+    newArray = runRW# $ \stateRW -> newArray# size init stateRW 
+
+
+-- | writeMutArr should have evaluation forced
+writeMutArr :: MutableArray# RealWorld a -> Int -> a -> ()
+writeMutArr mutArr (I# ix) val =
```

Comment:
This is an unsafe function, it should be in the name somehow. A good way would be to have `Unsafe` as the module name. Like `Data.Linear.Array.Unsafe`. So as to avoid repeating it for every function. There is already such an `Unsafe` module in `linear-base`.